virtual_ship/expedition/verify_schedule.py

```diff
@@ -0,0 +1,132 @@
+        ]
+    ):
+        raise PlanningError(
+            "Each waypoint should be timed after all previous waypoints"
+        )
+
+    # check if all waypoints are in water
+    # this is done by picking an arbitrary provided fieldset and checking if UV is not zero
+
+    # get all available fieldsets
+    available_fieldsets = [
+        fs
+        for fs in [
+            input_data.adcp_fieldset,
+            input_data.argo_float_fieldset,
+            input_data.ctd_fieldset,
+            input_data.drifter_fieldset,
+            input_data.ship_underwater_st_fieldset,
+        ]
+        if fs is not None
+    ]
+    # check if there are any fieldsets, else its an error
+    if len(available_fieldsets) == 0:
+        raise ValueError(
+            "No fieldsets provided to check if waypoints are on land. Assuming no provided fieldsets is an error."
+        )
+    # pick any
+    fieldset = available_fieldsets[0]
+    # get waypoints with 0 UV
+    land_waypoints = [
+        (wp_i, wp)
+        for wp_i, wp in enumerate(schedule.waypoints)
+        if _is_on_land_zero_uv(fieldset, wp)
+    ]
+    # raise an error if there are any
+    if len(land_waypoints) > 0:
+        raise PlanningError(
+            f"The following waypoints are on land: {['#' + str(wp_i) + ' ' + str(wp) for (wp_i, wp) in land_waypoints]}"
+        )
+
+    # check that ship will arrive on time at each waypoint (in case no unexpected event happen)
+    time = schedule.waypoints[0].time
+    for wp_i, (wp, wp_next) in enumerate(
+        zip(schedule.waypoints, schedule.waypoints[1:])
+    ):
+        if wp.instrument is InstrumentType.CTD:
+            time += timedelta(minutes=20)
+
+        geodinv: tuple[float, float, float] = projection.inv(
+            wp.location.lon, wp.location.lat, wp_next.location.lon, wp_next.location.lat
+        )
+        distance = geodinv[2]
+
+        time_to_reach = timedelta(
+            seconds=distance / ship_config.ship_speed_meter_per_second
+        )
+        arrival_time = time + time_to_reach
+
+        if wp_next.time is None:
+            time = arrival_time
+        elif arrival_time > wp_next.time:
+            raise PlanningError(
+                f"Waypoint planning is not valid: would arrive too late at a waypoint number {wp_i + 1}. location: {wp_next.location} time: {wp_next.time} instrument: {wp_next.instrument}"
+            )
+        else:
+            time = wp_next.time
+
+
+class PlanningError(RuntimeError):
+    """An error in the schedule."""
+
+    pass
+
+
+def _is_on_land_zero_uv(fieldset: FieldSet, waypoint: Waypoint) -> bool:
+    """
+    Check if waypoint is on land by assuming zero velocity means land.
+
+    :param fieldset: The fieldset to sample the velocity from.
+    :param waypoint: The waypoint to check.
+    :returns: If the waypoint is on land.
+    """
+    return fieldset.UV.eval(
+        0,
+        -fieldset.U.depth[0],
```

Comment:
The minus sign is a bit dangerous here; if the grid has not flipped. Perhaps add an `abs()` to force it to always be negative?

Reply:
improved sampling with new method. but to me it is strange that while we flip the depth, this way of obtaining the depth remains unflipped?